pulser/waveforms.py

```diff
@@ -233,6 +269,39 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f'CompositeWaveform({self.duration} ns, {self._waveforms!r})'
 
+    def __getitem__(self, index_or_slice:
+                    Union[int, slice]) -> Union[float, np.ndarray]:
+        wf_start: int
+        if isinstance(index_or_slice, slice):
+            s: slice = self._check_slice(index_or_slice)
+            start: int = s.start
+            stop: int = s.stop
+            array: np.ndarray = np.array([])
+            wf_start = 0
+            for wf in self._waveforms:
+                if wf_start <= start and start <= wf_start+wf.duration:
```

Comment:
Btw, I think you should start this block by checking if the start index is on a subsequent waveform, that way you can just skip to the next iteration of the loop:
```suggestion
                if start > wf_start + duration:
                    wf_start += duration
                    continue
                elif wf_start <= start:
```
Note how you can skip checking `start <= wf_start+wf.duration` too with the `elif`